File: 2nd Year/2nd Semester/MPP/PROIECT-JAVA/Laborator/TriatlonPersistence/src/main/java/triatlon/persistence/ProbaRepository.java
package triatlon.persistence;

import triatlon.model.Arbitru;
import triatlon.model.Proba;
import triatlon.model.TipProba;
import triatlon.persistence.JdbcUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;


public class ProbaRepository implements IProbaRepository {

    private static final Logger logger = LogManager.getLogger();
    private final JdbcUtils jdbcUtils;
    private final IArbitruRepository arbitruRepository;

    public ProbaRepository(Properties properties, IArbitruRepository arbitruRepository) {
        logger.info("Initializing ProbaRepository with proprieties: {}", properties);
        jdbcUtils = new JdbcUtils(properties);
        this.arbitruRepository = arbitruRepository;
    }

    @Override
    public Proba findOne(Integer integer) {
        logger.traceEntry("finding Proba with id {}", integer);
        String sql = "SELECT * FROM Proba WHERE id = ?";

        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql)) {

            preparedStatement.setInt(1, integer);

            try (ResultSet rs = preparedStatement.executeQuery()) {
                if (rs.next()) {
                    int arbitruId = rs.getInt("arbitru_id");
                    Arbitru arbitru = arbitruRepository.findOne(arbitruId);

                    String tipProbaStr = rs.getString("tip_proba");
                    TipProba tipProba = TipProba.valueOf(tipProbaStr);

                    return new Proba(
                            rs.getInt("id"),
                            tipProba,
                            arbitru
                    );
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("SQLException: " + e.getMessage());
        }
        return null;
    }

    @Override
    public Iterable<Proba> findAll() {
        List<Proba> probe = new ArrayList<>();
        logger.traceEntry("finding all Probe");
        String sql = "SELECT * FROM Proba";

        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql);
             ResultSet resultSet = preparedStatement.executeQuery()) {

            while (resultSet.next()) {
                int arbitruId = resultSet.getInt("arbitru_id");
                Arbitru arbitru = arbitruRepository.findOne(arbitruId);

                String tipProbaStr = resultSet.getString("tip_proba");
                TipProba tipProba = TipProba.valueOf(tipProbaStr);

                probe.add(new Proba(
                        resultSet.getInt("id"),
                        tipProba,
                        arbitru
                ));
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("SQLException: " + e.getMessage());
        }
        logger.traceExit();
        return probe;
    }

    @Override
    public void save(Proba entity) {
        logger.traceEntry("saving Proba {}", entity);
        String sql = "insert into Proba (tip_proba, arbitru_id) values (?, ?)";

        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql, PreparedStatement.RETURN_GENERATED_KEYS)) {

            preparedStatement.setString(1, entity.getTipProba().name());
            preparedStatement.setInt(2, entity.getArbitru().getId());

            int affectedRows = preparedStatement.executeUpdate();
            if (affectedRows > 0) {
                try (ResultSet generateKeys = preparedStatement.getGeneratedKeys()) {
                    if (generateKeys.next()) {
                        entity.setId(generateKeys.getInt(1));
                    }
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("SQLException: " + e.getMessage());
        }
        logger.traceExit();
    }

    @Override
    public void delete(Integer integer) {
        logger.traceEntry("delete Proba with id {}", integer);
        String sql = "delete from Proba where id = ?";
        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql)) {

            preparedStatement.setInt(1, integer);
            preparedStatement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("SQLException: " + e.getMessage());
        }
        logger.traceExit();
    }

    @Override
    public void update(Proba entity) {
        logger.traceEntry("update Proba {}", entity);
        String sql = "update Proba set tip_proba = ?, arbitru_id = ? where id = ?";

        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql)) {

            preparedStatement.setString(1, entity.getTipProba().name());
            preparedStatement.setInt(2, entity.getArbitru().getId());
            preparedStatement.setInt(3, entity.getId());

            preparedStatement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("SQLException: " + e.getMessage());
        }
        logger.traceExit();
    }
}

File: 2nd Year/2nd Semester/MPP/Triatlon-C#/TriatlonPersistance/ProbaRepository.cs
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Data.SQLite;
using TriatlonModel;
using log4net;

namespace TriatlonPersistance
{
    public class ProbaRepository : IProbaRepository
    {
        private static readonly ILog logger = LogManager.GetLogger(typeof(ProbaRepository));
        public readonly string connectionString;

        public ProbaRepository()
        {
            connectionString = ConfigurationManager.ConnectionStrings["DefaultConnection"].ConnectionString;
        }
        public void Delete(int id)
        {
            logger.Debug("Enter deleting Proba with id: " + id);
            string sql = "DELETE FROM Proba WHERE id=@id";
            try
            {
                using (SQLiteConnection connection = new SQLiteConnection(connectionString))
                {
                    connection.Open();
                    using (SQLiteCommand command = new SQLiteCommand(sql, connection))
                    {
                        command.Parameters.AddWithValue("@id", id);
                        command.ExecuteNonQuery();
                        logger.Debug("Deleted Proba with id: " + id);
                    }
                    connection.Close();
                }
                logger.Debug("Exit deleting Proba with id: " + id);
            }
            catch (SQLiteException ex)
            {
                logger.Error("Error deleting Proba with id: " + id, ex);
            }
        }

        public List<Proba> FindAll()
        {
            logger.Debug("Enter FindAll Proba");
            List<Proba> probe = new List<Proba>();
            string sql = "SELECT * FROM Proba";
            try
            {
                using (SQLiteConnection connection = new SQLiteConnection(connectionString))
                {
                    connection.Open();
                    using (SQLiteCommand command = new SQLiteCommand(sql, connection))
                    {
                        using (SQLiteDataReader reader = command.ExecuteReader())
                        {
                            while (reader.Read())
                            {
                                int arbitru_id = reader.GetInt32(2);
                                Arbitru arbitru = new ArbitruRepository().FindOne(arbitru_id);
                                String tipProbaStr = reader.GetString(1);
                                TipProba tipProba = (TipProba)Enum.Parse(typeof(TipProba), tipProbaStr);
                                Proba proba = new Proba(reader.GetInt32(0), tipProba, arbitru);
                                probe.Add(proba);
                                logger.Debug("Found Proba: " + proba);
                            }
                        }
                    }
                    connection.Close();
                }
            }
            catch (SQLiteException ex)
            {
                logger.Error("Error FindAll Proba", ex);
            }
            logger.Debug("Exit FindAll Proba");
            return probe;
        }

        public Proba FindOne(int id)
        {
            logger.Debug("Enter FindOne Proba with id: " + id);
            string sql = "SELECT * FROM Proba WHERE id=@id";
            try
            {
                using (SQLiteConnection connection = new SQLiteConnection(connectionString))
                {
                    connection.Open();
                    using (SQLiteCommand command = new SQLiteCommand(sql, connection))
                    {
                        command.Parameters.AddWithValue("@id", id);
                        using (SQLiteDataReader reader = command.ExecuteReader())
                        {
                            if (reader.Read())
                            {
                                int arbitru_id = reader.GetInt32(2);
                                Arbitru arbitru = new ArbitruRepository().FindOne(arbitru_id);
                                String tipProbaStr = reader.GetString(1);
                                TipProba tipProba = (TipProba)Enum.Parse(typeof(TipProba), tipProbaStr);
                                Proba proba = new Proba(reader.GetInt32(0), tipProba, arbitru);
                                logger.Debug("Found Proba: " + proba);
                                return proba;
                            }
                        }
                    }
                    connection.Close();
                }
            }
            catch (SQLiteException ex)
            {
                logger.Error("Error FindOne Proba with id: " + id, ex);
            }
            return null;
        }

        public void Save(Proba entity)
        {
            logger.Debug("Saving Proba " + entity);
            string sql = "INSERT INTO Proba (tip_proba, arbitru_id) VALUES (@TipProba, @ArbitruId)";
            try
            {
                using (SQLiteConnection connection = new SQLiteConnection(connectionString))
                {
                    connection.Open();
                    using (SQLiteCommand command = new SQLiteCommand(sql, connection))
                    {

                        command.Parameters.AddWithValue("@TipProba", entity.TipProba.ToString());
                        command.Parameters.AddWithValue("@ArbitruId", entity.Arbitru.Id);

                        int affectedRows = command.ExecuteNonQuery();
                        if (affectedRows > 0)
                        {
                            command.CommandText = "SELECT last_insert_rowid()";
                            entity.Id = Convert.ToInt32(command.ExecuteScalar());
                        }
                        logger.Debug("Saved Proba " + entity);
                    }
                    connection.Close();
                }
            }
            catch (SQLiteException ex)
            {
                logger.Error("Error saving Proba " + entity, ex);
            }
            logger.Debug("Exit Save Proba " + entity);
        }

        public void Update(Proba entity)
        {
            logger.Debug("Enter Update Proba " + entity);
            string sql = "UPDATE Proba SET tip_proba=@TipProba, arbitru_id=@ArbitruId WHERE id=@id";
            try
            {
                using (SQLiteConnection connection = new SQLiteConnection(connectionString))
                {
                    connection.Open();
                    using (SQLiteCommand command = new SQLiteCommand(sql, connection))
                    {
                        command.Parameters.AddWithValue("@TipProba", entity.TipProba.ToString());
                        command.Parameters.AddWithValue("@ArbitruId", entity.Arbitru.Id);
                        command.Parameters.AddWithValue("@id", entity.Id);
                        command.ExecuteNonQuery();
                        logger.Debug("Updated Proba " + entity);
                    }
                    connection.Close();
                }
            }
            catch (SQLiteException ex)
            {
                logger.Error("Error updating Proba " + entity, ex);
            }
        }
    }
}
